### PR TITLE
CIRCSTORE-329 Implement TLR feature toggle job service

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose 
- Check TLR feature status;
- If it's enabled, recalculate and update a position field according to the "unified queue" rules (one queue per instance);
- If it's disabled, recalculate and update a position field according to the "old" rules - one queue per item;
## Approach
- implement the logic for TLR feature toggle job service;
- create tests;

Resolves: [CIRCSTORE-329](https://issues.folio.org/browse/CIRCSTORE-329)

### CODE REVIEW

File: src/test/java/org/folio/rest/api/TlrFeatureToggleJobAPITest.java [hunk 1]
```diff
@@ -138,6 +158,109 @@ public void processingShouldRespondWith202WhenThereAreRunningJobs() throws Malfo
...
+    assertThat(jobPostResponse.getStatusCode(), is(202));
+    Response updatedJob = getTlrFeatureToggleJobById(postResponse.getJson().getString("id"));
+    assertThat(updatedJob.getJson().getString("status"), is(IN_PROGRESS.toString()));
+
+    assertRequestsQueue();
+  }
+
+  private void assertRequestsQueue() throws MalformedURLException, ExecutionException,
+    InterruptedException, TimeoutException {
+
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/api/TlrFeatureToggleJobAPITest.java:201]: Group requests by instanceId, check their positions within the group

File: src/test/java/org/folio/rest/api/TlrFeatureToggleJobAPITest.java [hunk 2]
```diff
@@ -138,6 +164,137 @@ public void processingShouldRespondWith202WhenThereAreRunningJobs() throws Malfo
...
+      .collect(Collectors.toList());
+    assertThat(secondInstancePositions, hasSize(2));
+    assertThat(secondInstancePositions.containsAll(List.of("1", "2")), is(true));
+
+    Response updatedJob = getTlrFeatureToggleJobById(postResponse.getJson().getString("id"));
+    await().until(() -> updatedJob.getJson().getString("status"), is(DONE.toString()));
+  }
+
+  private void createTitleLevelRequestsQueue(UUID firstInstanceId, UUID secondInstanceId) {
+    String requestLevel = "Title";
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/api/TlrFeatureToggleJobAPITest.java:246]: Item

File: src/test/java/org/folio/rest/api/TlrFeatureToggleJobAPITest.java [hunk 3]
```diff
@@ -138,6 +164,137 @@ public void processingShouldRespondWith202WhenThereAreRunningJobs() throws Malfo
...
+    assertThat(getResponse.getStatusCode(), is(200));
+    List<JsonObject> updatedRequests = JsonArrayHelper.toList(getResponse.getJson()
+      .getJsonArray("requests"));
+    assertThat(updatedRequests, hasSize(7));
+    List<String> firstInstancePositions = updatedRequests.stream()
+      .filter(request -> firstInstanceId.toString().equals(request.getString("instanceId")))
+      .map(request -> request.getString("position"))
+      .collect(Collectors.toList());
+    assertThat(firstInstancePositions, hasSize(5));
+    assertThat(firstInstancePositions.containsAll(List.of("1", "2", "3", "4", "5")), is(true));
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/api/TlrFeatureToggleJobAPITest.java:193]: make it more strict

**Review by alexanderkurash — APPROVED** [review]

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java [hunk 4]
```diff
@@ -1,26 +1,151 @@
...
+    return updatedRequests;
+  }
+
+  private Future<List<TlrFeatureToggleJob>> findJobsByStatus(String status) {
+    return tlrFeatureToggleJobRepository.get(new Criterion()
+      .addCriterion(new Criteria()
+        .addField(TLR_FEATURE_TOGGLE_JOB_STATUS_FIELD)
+        .setOperation("=")
+        .setVal(status))
+      .setOrder(new Order(String.format("jsonb->>'%s'","metadata,createdDate"),
```
Comment (alexanderkurash) [inline src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java:141]: space

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java [hunk 5]
```diff
@@ -1,26 +1,151 @@
...
+  private List<Request> updatePosition(Map<String, List<Request>> groupedRequests) {
+    List<Request> updatedRequests = new ArrayList<>();
+      for (var requests : groupedRequests.values()) {
+        requests.sort(Comparator.comparingInt(Request::getPosition));
+        int position = 1;
+        for (var request : requests) {
+          request.setPosition(position++);
+        }
+        updatedRequests.addAll(requests);
+      }
```
Comment (alexanderkurash) [inline src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java:130]: Indentation

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java [hunk 6]
```diff
@@ -1,26 +1,151 @@
...
+
```
Comment (alexanderkurash) [inline src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java:131]: Set number of processed requests

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java (hunk 4 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java]: done

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java (hunk 5 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java]: done

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java [hunk 7]
```diff
@@ -1,26 +1,151 @@
...
+          .addField(REQUEST_STATUS_FIELD)
+          .setOperation("=")
+          .setVal(status.toString())))
+      .flatMap(Collection::stream)
+      .collect(toList());
+  }
+
+  private GroupedCriterias buildGroupedCriterias(List<Criteria> criterias) {
+    GroupedCriterias groupedCriterias = new GroupedCriterias();
+    criterias.forEach(criteria -> groupedCriterias.addCriteria(criteria, "OR"));
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java:117]: done

File: src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java (hunk 6 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/service/tlr/TlrFeatureToggleService.java:135]: done

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**